### DISCUSSION TITLE
Markdoc for API documentation

### DISCUSSION THREAD

**Discussion by PrabhavSF** [markdoc/markdoc#192] [Q&A] [answered]
Can we use Markdoc for docs generated straight from OAS specs? Are the Stripe API docs generated using Markdoc?

**Comment by mfix-stripe** [comment, marked as answer]
@PrabhavSF yes! Markdoc is a low enough level toolchain that you can create or generate documents from any source you like, including OpenAPI specs. We are currently exploring ways to generate the scaffolding of our API docs from our spec.

**Reply by Patriksafar** [reply]
Hey @mfix-stripe I know there are some tools like swagger which could do that, but I wanted to ask if you plan to support that in near future by default. 

EDIT: For now I have came up with custom node script which generates the docs from Open api schema - not the perfect solution but for now it works for us.

**Reply by nirgilboa** [reply]
hey @mfix-stripe is there anything new with generating docs from OpenAPI specs? 🙏🏻 I'm evaluating Markdoc and wondering whether we could include an endpoint or tag within the spec by pointing to it. Obviously we could develop a custom tag for this purpose, but I wonder if something exists already

**Reply by adriangalilea** [reply]
@mfix-stripe bump

**Reply by mfix-stripe** [reply]
@nirgilboa @adriangalilea creating a custom tag for this is the way to go! Our own API docs are generated by taking our OpenAPI spec, doing some customizations, creating a Markdoc AST, and then generating a bunch of Markdoc files from each AST. This solution is entirely custom to our infra though, and there isn't much to pull from. 

We don't plan on releasing solutions like this, but rather focus on building out the low-level toolchain that is Markdoc, and letting the ecosystem build business-level solutions on top of it.

**Reply by gkpo** [reply]
> Hey @mfix-stripe I know there are some tools like swagger which could do that, but I wanted to ask if you plan to support that in near future by default.
> 
> EDIT: For now I have came up with custom node script which generates the docs from Open api schema - not the perfect solution but for now it works for us.

@Patriksafar 
Would be really interested to see how your script works! :) trying to implement something similar on our side

**Reply by mfix-stripe** [reply]
@gkpo we have an entire system dedicated to this—it wouldn't be as simple as sharing a script to demonstrate what I mean, sorry. 

We have not documented `Markdoc.format` yet, which is used for the "generating a bunch of Markdoc files" step, but hope to shortly.

**Reply by gkpo** [reply]
@mfix-stripe  I understand thanks for your reply nonetheless :) what is the main idea of your script? parse the openapi.json file, look for endpoints and go from there?

**Reply by mfix-stripe** [reply]
@gkpo exactly—endpoints, parameters, etc. As for the final step, just added documentation for `Markdoc.format` which is used to generate the source file: https://markdoc.dev/docs/format

**Reply by Patriksafar** [reply]
@gkpo For sure you would find better solution but since you asked:

A year ago I created script which parse the openapi schema, look for endpoints and creating markdown page for each endpoint with react components.

Schema is located at: `pages/${inputPath}/api/schema/schema.yml`  and all pages are created to `./pages/${inputPath}/api/`

Just run `node scripts/generate-api-pages.js --input='doc-api'`

```js
// scripts/generate-api-pages.js

const fs = require('fs')

const yaml = require('yaml')

// Access command-line arguments
// eslint-disable-next-line @typescript-eslint/no-magic-numbers
const args = process.argv.slice(2)
const inputParam = args.find((arg) => arg.startsWith('--input='))

// Extract the value from the --path argument
let inputPath = ''

if (inputParam !== undefined) {
  inputPath = inputParam.split('=')[1]
}

const schemaData = fs.readFileSync(
  `./pages/${inputPath}/api/schema/schema.yml`,
  'utf-8'
)
const schema = yaml.parse(schemaData)

function replaceRefBySchemaComponent(refValue) {
  if (refValue === undefined) {
    return
  }
  const componentName = refValue.split('/').pop()
  const component = schema.components.schemas[componentName]

  return component
}

function resolveReferences(obj) {
  /* eslint-disable no-param-reassign */
  if (typeof obj === 'object' && obj !== null) {
    if (Array.isArray(obj)) {
      obj.forEach((item, index) => {
        obj[index] = resolveReferences(item)
      })
    } else {
      for (const key in obj) {
        if (key === '$ref') {
          const ref = obj[key]
          delete obj[key]
          obj.response = replaceRefBySchemaComponent(ref)
        } else {
          obj[key] = resolveReferences(obj[key])
        }
      }
    }
  }

  return obj
  /* eslint-enable no-param-reassign */
}

const outputDirectory = `./pages/${inputPath}/api/`

// Ensure the output directory exists
fs.mkdirSync(outputDirectory, { recursive: true })

function generatePageComponent(
  path,
  method,
  operation,
  pageName,
  securitySchemes
) {
  const formattedOperation = operation
  const pageTitle = operation.summary ?? operation.operationId
  const pageDescription = operation.description ?? ''
  const displayBearerAuthInfo =
    securitySchemes?.SMIPaymentProcessorKeyAuth?.scheme === 'bearer' ||
    securitySchemes?.EshopTokenAuth?.scheme === 'bearer' ||
    securitySchemes?.PaymentProcessorKeyAuth?.scheme === 'token' ||
    false

  const displayTokenAuthInfo =
    securitySchemes?.PaymentProcessorKeyAuth?.type === 'apiKey' || false

  const pageMarkdown = `---
title: ${pageTitle}
section: ${inputPath}
---

<!-- This page is autogenerated - DO NOT EDIT IT DIRECTLY! -->

# ${pageTitle}

{% fetchEndpointBox
    method="${method}"
    path="${schema.servers[0].url}"
    endpoint="${path}"
/%}

${pageDescription}

${
  displayBearerAuthInfo
    ? `
  {% banner variant="warning" %}
    **Security: Bearer Auth**

    Provide your bearer token in the Authorization header when making requests to protected resources.  
    Example: \`Authorization: Bearer 123\`
  {% /banner %}
`
    : displayTokenAuthInfo
    ? `
    {% banner variant="warning" %}
    **Security: Token Auth**

    Provide your auth token in the Authorization header when making requests to protected resources.  
    Example: \`Authorization: Token 123\`
  {% /banner %}
`
    : ''
}

{% apiView
  apiSchemaContent=${JSON.stringify(formattedOperation).replace(/\\n\\n/g, '')}
/%}
`

  fs.writeFileSync(`${outputDirectory}${pageName}.md`, pageMarkdown)
}

function generateNextJsPages(schema) {
  const paths = schema.paths
  const securitySchemes = schema.components?.securitySchemes ?? {}
  const links = []

  for (const [path, methods] of Object.entries(paths)) {
    for (const [method, operation] of Object.entries(methods)) {
      const pageName =
        operation.summary !== null && operation.summary !== undefined
          ? operation.summary
          : operation.operationId

      const formattedPageName = pageName
        .replace(/([a-z])([A-Z])/g, '$1-$2')
        .toLowerCase() // add `-` as word separator

      // generate page file
      generatePageComponent(
        path,
        method,
        operation,
        formattedPageName,
        securitySchemes
      )

      // push link to this page to the links array
      links.push({
        children: `${pageName} ${method.toUpperCase()}`,
        href: `/${inputPath}/api/${formattedPageName}`,
      })
    }
  }

  const linkArrayContent = `export const ${inputPath.replace(
    '-',
    ''
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
  )}Navigation = ${JSON.stringify(links, null, 2)};\n`

  fs.writeFileSync(`config/${inputPath}-navigation.ts`, linkArrayContent)
}

// remove all previous files from the output directory but keep only 'schema' directory
fs.readdirSync(outputDirectory)
  .filter((directory) => directory !== 'schema')
  .forEach((file) => fs.unlinkSync(`${outputDirectory}/${file}`))

generateNextJsPages(resolveReferences(schema))
```

**Reply by anthonyvallone-paydock** [reply]
@Patriksafar thats a create solution. Thanks for sharing.
